### DISCUSSION TITLE
Should the course modules be more connected?

### DISCUSSION THREAD

**Discussion by edwardchalstrey1** [alan-turing-institute/rsd-engineeringcourse#168] [Ideas]
I think there could be something to be gained from threading the modules together more than they were currently. Perhaps converting some of the exercises into parts of a larger “project” that the students work on throughout the course, so by the middle of the course they have a GitHub repo set up with their code and tests running in CI, and by the end of the course they have added to and modified this same codebase multiple times, done refactoring, added new modules that address different functionality (other exercises) ultimately releasing a packaged version to PyPI or something like that

Having taught the version control and testing modules back to back, I found it useful to contextualise running tests by saying things like "we commit the code, but the tests fail, so we make another change that allows the test to pass, then we commit this change". This is what prompted this suggestion.

@jack89roberts comments that this is how the RDS course works. However, for the RSE course there was a conscious decision a couple of years ago to try to make the modules independent, so students didn’t necessarily have to attend the whole course (and less of a problem if someone didn’t finish the exercises the day before).

Having said that, I think should be possible for course developers to be careful to make it so that “falling behind” wasn’t too much of an issue, if the parts of the "project" were not dependent on one another too much.

My suggestion could make it harder for the instructors if we continue to split the teaching as we have this year. If we did get applied skills service area budget to spend some more FTE on improving the course, perhaps people assigned to do so could also be the teachers (assuming there are people in REG that would be keen to teach a 2 week course, which there may not be).

Thoughts?

**Comment by jack89roberts** [comment]
Another comment I have is that creating exercises, especially one that is meaty enough to work for a big chunk of the course, takes a lot of time and effort. We struggle to maintain/iteratively improve the course with the current level of effort.

I do think it's a nice idea though, especially if we were to dedicate more time to hands-on sessions/exercises (also like the RDS course). Having sample solutions/snapshots of the repo and how it should look after each module would get around some of the problems with falling behind/having independent modules.

**Comment by jemrobinson** [comment]
I do like the idea behind this, but I see a couple of potential problems with it:

1. People who miss a module being unable to work on the next module
2. People who didn't finish the exercises being left behind on the next day

I think I'd actually prefer to see smaller exercises that can be done in 5 minutes or less with a solution quickly covered during the course.

**Reply by edwardchalstrey1** [reply]
yes, I definitely agree with smaller exercises
